src/Traits/VersionsCacheTrait.php

```diff
@@ -0,0 +1,187 @@
+use SilverStripe\Model\List\ArrayList;
+use SilverStripe\ORM\Connect\Query;
+use SilverStripe\Versioned\Versioned;
+use SilverStripe\Versioned\Versioned_Version;
+use Stringable;
+use SilverStripe\ORM\DataList;
+use SilverStripe\ORM\Queries\SQLConditionGroup;
+
+/**
+ * Functionality to cache Versioned::Versions(), which is used for the <DataObject>_Versions tables
+ * This trait exists to help encapsulate specific functionality for the Versioned class.
+ * Its use with any other class is not supported.
+ * @internal
+ */
+trait VersionsCacheTrait
+{
+    /**
+     * In-memory cache of Versions i.e. the <DataObject>_Versions table
+     * @internal
+     */
+    private static array $versions_cache = [];
+
+    /**
+     * Determines whether to use cached data in the Versions() method
+     * @internal
+     */
+    private static bool $useVersionsCache = false;
+
+    /**
+     * Get whether to use cached data in the Versions() method
+     */
+    public static function getUseVersionsCache(): bool
+    {
+        return self::class::$useVersionsCache;
+    }
+
+    /**
+     * Set whether to use cached data in the Versions() method
+     */
+    public static function setUseVersionsCache(bool $use): void
+    {
+        self::class::$useVersionsCache = $use;
+    }
+
+    /**
+     * Prepopulate the in-memory Versions cache, which are used for <DataObject>_Versions tables
+     * The values will be retrieved using an efficient `WHERE "ID" IN (<ids>)` SQL query
+     * Note method this should not to be confused with prepopulateVersionNumberCache()
+     */
+    public static function prepopulateVersionsCache(string $baseDataClass, array $ids): void
+    {
+        $cachedIDs = [];
+        foreach ($ids as $id) {
+            $key = self::class::generateVersionsCacheKey($id);
+            if (array_key_exists($key, self::class::$versions_cache)) {
+                $cachedIDs[] = $id;
+            }
+        }
+        // Only fetch uncached IDs
+        $ids = array_diff($ids, $cachedIDs);
+        if (empty($ids)) {
+            return;
+        }
+        $versions = self::class::innerGetVersions($baseDataClass, $ids);
+        foreach ($ids as $id) {
+            $key = self::class::generateVersionsCacheKey($id);
+            if (array_key_exists($key, self::class::$versions_cache)) {
+                continue;
+            }
+            $vers = array_filter($versions, fn($ver) => (int) $ver->RecordID === $id);
+            self::class::$versions_cache[$key] = ArrayList::create($vers);
+        }
+    }
+
+    /**
+     * Get an ArrayList of Versions for an ID
+     * Will use a cache if setUseVersionsCache(true) was previously called
+     *
+     * @return ArrayList<Versioned_Version>
+     */
+    private static function readVersionsByID(
+        DataObject $owner,
+        string $baseDataClass,
+        int $id,
+        mixed $filter,
+        mixed $sort,
+        mixed $limit,
+        mixed $join,
+    ): ArrayList {
```

Comment:
```suggestion
     * @return SSList<Versioned_Version>
     */
    private static function readVersionsByID(
        DataObject $owner,
        string $baseDataClass,
        int $id,
        mixed $filter,
        mixed $sort,
        mixed $limit,
        mixed $join,
    ): SSList {
```


Sorry, should have caught this last review. Lets just say it'll be any list, which lets us use a `DataList` if we find a good way to do so in the future.
I know we could update that at the same time since it's private, but may as well do now when we add it in the first place (and if we promote visibility later we won't have to worry about it).

Reply:
I really don't agree with this, it's a private method, and we'll probably never actually update to support more things.

More importantly, it's used directly by Versioned::Versions(), which has a docblocked return type of ArrayList